### PR TITLE
Made exceptions caught by get_component_class more specific

### PR DESCRIPTION
**Pull request by hsm207**
**Proposed changes**:
- Made exceptions caught by get_component_class more specific. See issue #1513 for more details.
- closes https://github.com/RasaHQ/rasa/issues/1513

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

**Comment by hsm207** [comment]
What should I change to pass the tests in Python 3.5?

**Comment by msamogh** [comment]
Thanks for the PR, we'll give it a review as soon as possible.

File: rasa/nlu/registry.py [hunk 1]
```diff
@@ -139,15 +139,29 @@ def get_component_class(component_name: Text) -> Type["Component"]:
...
+                # when component_name is a path to a class but that path is invalid
                 raise Exception(
-                    "Failed to find component class for '{}'. Unknown "
-                    "component name. Check your configured pipeline and make "
-                    "sure the mentioned component is not misspelled. If you "
-                    "are creating your own component, make sure it is either "
-                    "listed as part of the `component_classes` in "
-                    "`rasa.nlu.registry.py` or is a proper name of a class "
-                    "in a module.".format(component_name)
+                    "Failed to find component class for '{}'.Unknown component name.\n{}".format(
```
Comment (wochinge) [inline rasa/nlu/registry.py]:
```suggestion
                    "Failed to find module '{}'. \n{}".format(
```

File: rasa/nlu/registry.py [hunk 2]
```diff
@@ -139,15 +139,29 @@ def get_component_class(component_name: Text) -> Type["Component"]:
...
+                        component_name, e.msg
+                    )
+                )
+            except AttributeError:
+                # when component_name is a path to a class but the path does not contain that class
+                module_name, _, class_name = component_name.rpartition(".")
+                raise Exception(
+                    "Failed to find component class for '{}'.Unknown component name.\n"
```
Comment (wochinge) [inline rasa/nlu/registry.py]:
```suggestion
                    "Failed to find class '{}' in module '{}'.\n"
```

File: rasa/nlu/registry.py [hunk 3]
```diff
@@ -139,15 +139,29 @@ def get_component_class(component_name: Text) -> Type["Component"]:
...
+                    "Cannot find class '{}' in module {}.".format(
```
Comment (wochinge) [inline rasa/nlu/registry.py]: I would shorten the messages a bit.
```suggestion
                    "".format(
```

File: rasa/nlu/registry.py [hunk 4]
```diff
@@ -139,15 +139,29 @@ def get_component_class(component_name: Text) -> Type["Component"]:
...
+                        component_name, class_name, module_name
+                    )
+                )
+            except ImportError:
+                # when component_name is a class name and not part of old_style_names
+                raise Exception(
+                    "Failed to find component class for '{0}'.Unknown component name.\n"
```
Comment (wochinge) [inline rasa/nlu/registry.py]: I think it's okay if we just keep the second part. Can we provide any solution in the exception message or a link to some explanations?
```suggestion
                    "Cannot import class '{0}' from global namespace.\n"
```

File: tests/nlu/base/test_components.py [hunk 5]
```diff
@@ -58,7 +58,21 @@ def test_builder_create_unknown(component_builder, default_config):
     with pytest.raises(Exception) as excinfo:
         component_config = {"name": "my_made_up_componment"}
         component_builder.create_component(component_config, default_config)
-    assert "Unknown component name" in str(excinfo.value)
+    assert "Cannot import class" in str(excinfo.value)
+
+
+def test_builder_create_by_module_path_wrong_class(component_builder, default_config):
```
Comment (wochinge) [inline tests/nlu/base/test_components.py]: can we use this https://docs.pytest.org/en/latest/parametrize.html to parametrize the tests for the different cases?

File: rasa/nlu/registry.py (hunk 4 above)
Comment (hsm207) [reply rasa/nlu/registry.py]: What about something like?

`"Cannot import class '{0}' from global namespace.\nCheck that you have not made a typo in the class name and that you have installed the correct package."
`

Comment (wochinge) [reply rasa/nlu/registry.py]: Sounds good!

**Comment by hsm207** [comment]
@wochinge sorry for the delay. I've used pytest to parameterize unit tests before so I don't think I will encounter any issues there. It's just that I'm on a long business trip so I haven't had the time to code this up. I'll try completing it by the end of next week.

File: tests/nlu/base/test_components.py (hunk 5 above)
Comment (hsm207) [reply tests/nlu/base/test_components.py]: @wochinge Done. Please let me know if you have any feedback.

File: rasa/nlu/registry.py (hunk 4 above)
Comment (hsm207) [reply rasa/nlu/registry.py]: @wochinge I don't think the explanation is correct. 

That exception is triggered when the class cannot be retrieved from globals i.e.

`module = globals().get(module_path, locals().get(module_path))`

fails and where `module_path` is the class name.


So I think the message should be:

`Cannot find class '{0}' from global namespace. Check that you have not made a typo in the class name and that you have imported the class into the global namespace.`

But I am not sure at what point the user will have a chance to manually import the class into the global namespace.

Thoughts?

Also, the unit tests I've seen so far has `lookup_path` set to `None`.  Do you know what are the circumstances when `lookup_path` is not `None`?

**Comment by hsm207** [comment]
> @hsm207 would be cool if we could merge this soon cause that's a very helpful addition. Can I help you somehow with the last bits and pieces?

I've parameterized the tests.

File: rasa/nlu/registry.py (hunk 4 above)
Comment (wochinge) [reply rasa/nlu/registry.py]: > Also, the unit tests I've seen so far has lookup_path set to None. Do you know what are the circumstances when lookup_path is not None?

which `lookup_path`? 

> Cannot find class '{0}' from global namespace. Check that you have not made a typo in the class name and that you have imported the class into the global namespace.

Ok 👍 
Let's do `Cannot find class '{0}' from global namespace. Please check that there is no typo in the class name and that you have imported the class into the global namespace.

Comment (hsm207) [reply rasa/nlu/registry.py]: > which lookup_path?

I'm referring to the `lookup_path` in [this ](https://github.com/RasaHQ/rasa/blob/e5f2375d3916fdc4ab568ac45e03f9c67bd2f137/rasa/utils/common.py#L163-L187)function's argument.

> Ok 👍
> Let's do `Cannot find class '{0}' from global namespace. Please check that there is no typo in the class name and that you have imported the class into the global namespace.

Done!

**Comment by hsm207** [comment]
Done!

File: rasa/nlu/registry.py [hunk 6]
```diff
@@ -140,15 +140,25 @@ def get_component_class(component_name: Text) -> Type["Component"]:
...
+                # when component_name is a path to a class but the path does not contain that class
+                module_name, _, class_name = component_name.rpartition(".")
+                raise Exception(
+                    "Failed to find class '{}' in module '{}'.\n"
+                    "".format(component_name, class_name, module_name)
+                )
+            except ImportError:
+                # when component_name is a class name and not part of old_style_names
+                raise Exception(
+                    "Cannot find class '{0}' from global namespace. Please check that there is no typo in the class "
```
Comment (wochinge) [inline rasa/nlu/registry.py]:
```suggestion
                    "Cannot find class '{}' from global namespace. Please check that there is no typo in the class "
```

**Comment by wochinge** [comment]
@hsm207 Could you please also update the changelog? :-)

**Comment by hsm207** [comment]
@wochinge done. It's my first time updating a changelog. Please let me know if I've made any mistakes.

File: CHANGELOG.rst [hunk 7]
```diff
@@ -33,6 +33,8 @@ Changed
 - ``RasaChatInput`` fetches the public key from the Rasa X API. The key is used to
   decode the bearer token containing the conversation ID. This requires
   ``rasa-x>=0.20.2``.
+- more specific exception message from ``get_component_class()`` depending on whether component's path or
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
- more specific exception message when loading custom components depending on whether component's path or
```

**Comment by wochinge** [comment]
Made a tiny suggestion, but looks great @hsm207 ! 🎉

**Comment by hsm207** [comment]
> Made a tiny suggestion, but looks great @hsm207 ! 🎉

Suggestion committed!

**Comment by wochinge** [comment]
Cool, I merged in the latest master, then we wait for travis, and then we can merge it 🎉

**Comment by wochinge** [comment]
@hsm207 Can you please check why the tests fail?

**Comment by hsm207** [comment]
@hsm207 I don't understand what's going on.

The tests I wrote are failing for 3.5 but not for 3.6 and 3.7. Here's an example of the output of a failed test in 3.5:

image

Any ideas as to what might be causing the failures?

**Comment by wochinge** [comment]
@hsm207 `ModuleNotFoundError` was added in Python 3.6 so it's not available in Python 3.5: https://docs.python.org/3/library/exceptions.html#ModuleNotFoundError . You have to add a check for the Python version and in case it's Python 3.5 you have to skip the except `ModuleNotFoundError` (or maybe replace it with an `ImportError` 🤔 )

**Comment by wochinge** [comment]
@hsm207 I am on vacation for the next weeks, but @tabergma will take over for me in the meanwhile so that we can merge this :-)

**Comment by hsm207** [comment]
@wochinge @tabergma Could you please review my latest commit? It's a very hacky way of passing the unit tests for Python 3.5. Any suggestions to improve it?

@wochinge enjoy your vacation!

**Comment by tabergma** [comment]
I actually don't like that hack. Why not simply catching the `ImportError` and merge the two log messages? Would additionally add a comment to use `ModuleNotFoundError` as soon as we don't longer support python3.5

**Comment by hsm207** [comment]
@tabergma  So instead of this:

        if component_name not in old_style_names:
            try:
                return class_from_module_path(component_name)
            except ModuleNotFoundError as e:
                # when component_name is a path to a class but that path is invalid
                raise Exception(
                    "Failed to find module '{}'. \n{}".format(component_name, e.msg)
                )
            except AttributeError:
                # when component_name is a path to a class but the path does not contain that class
                module_name, _, class_name = component_name.rpartition(".")
                raise Exception(
                    "Failed to find class '{}' in module '{}'.\n"
                    "".format(component_name, class_name, module_name)
                )
            except ImportError:
                # when component_name is a class name and not part of old_style_names
                raise Exception(
                    "Cannot find class '{0}' from global namespace. Please check that there is no typo in the class "
                    "name and that you have imported the class into the global namespace.".format(
                        component_name
                    )
                )
        else:

Are you suggesting that we merge the exception message in `ModuleNotFoundError` into `ImportError` and only catch `AttributeError` and `ImportError`?

**Comment by tabergma** [comment]
Yes. What do you think?

**Comment by hsm207** [comment]
It's simple. I like it.

**Comment by tabergma** [comment]
Would be great if you can update the PR soon, so that we can merge it. Thank you!

**Comment by hsm207** [comment]
@tabergma PR updated. Please review and let me know if there are any changes I should make.

**Comment by federicotdn** [comment]
Hey @hsm207, all unit tests have passed but the Travis build is failing because of some style checks: https://travis-ci.com/RasaHQ/rasa/jobs/230005401. If you fix those, we can then merge your PR. You can run the style checks locally if you want, the instructions on how to do that can be found on README.md.

**Comment by hsm207** [comment]
@federicotdn done. Sorry about that.

**Comment by federicotdn** [comment]
Merged! 👍 
Thanks for your contribution @hsm207!!

**Pull request merged**